docker/Dockerfile

```diff
@@ -0,0 +1,244 @@
+    \
+&&  echo "## Done"
+
+# This generates configuration that contains all valid paths according to installed SDK
+RUN cd ${EMSDK} \
+    &&  echo "## Generate standard configuration" \
+    \
+    &&  ./emsdk activate ${EMSCRIPTEN_VERSION} --embedded \
+    &&  ./emsdk construct_env > /dev/null \
+    &&  cat ${EMSDK}/emsdk_set_env.sh \
+    \
+    # remove wrongly created entry with EM_CACHE, variable will be picked up from ENV
+    &&  sed -i -e "/EM_CACHE/d" ${EMSDK}/emsdk_set_env.sh \
+    # add a link to tools like asm2wasm in a system path
+    # asm2wasm (and friends might be places either in ./upstream of ./fastcomp folder, hence detection is needed)
+    &&  printf "export PATH=$(dirname $(find . -name asm2wasm -exec readlink -f {} +)):\$PATH\n" >> ${EMSDK}/emsdk_set_env.sh \
+    \
+&&  echo "## Done"
+
+# Create a structure and make mutable folders accessible for r/w
+RUN cd ${EMSDK} \
+    &&  echo "## Create .data structure" \
+    &&  for mutable_dir in ${EM_DATA} ${EM_PORTS} ${EM_CACHE} ${EMSDK}/zips ${EMSDK}/tmp; do \
+            mkdir -p ${mutable_dir}; \
+            chmod -R 777 ${mutable_dir}; \
+        done \
+        \
+&&  echo "## Done"
+
+# Create an entrypoint that activates Emscripten SDK and helps running this image as non-root user
+COPY entrypoint ${EMSDK}/
+
+# Internal test suite of tools that this image provides
+COPY test_dockerimage.sh ${EMSDK}/
+
+
+# Create symbolic links for critical Emscripten Tools
+# This is important for letting people using Emscripten in Dockerfiles without activation
+# As each Emscripten release is placed to a different folder (i.e. /emsdk_portable/emscripten/tag-1.38.31)
+RUN echo "## Create symbolic links" \
+    &&  . ${EMSDK}/emsdk_set_env.sh \
+    \
+    &&  mkdir -p ${EMSDK}/llvm ${EMSDK}/emscripten ${EMSDK}/binaryen \
+    \
+    &&  ln -s $(dirname $(which node))/..       ${EMSDK}/node/current \
+    &&  ln -s $(dirname $(which clang))/..      ${EMSDK}/llvm/clang \
+    &&  ln -s $(dirname $(which emcc))          ${EMSDK}/emscripten/sdk \
+    \
+    &&  ln -s $(dirname $(which asm2wasm))      ${EMSDK}/binaryen/bin \
+    \
+    &&  echo "## Done"
+
+# Clean up emscripten installation and strip some symbols
+RUN echo "## Aggresive optimization: Remove debug symbols" \
+&&  apt-get -qq -y update && apt-get -qq install -y --no-install-recommends \
+        binutils \
+    && . ${EMSDK}/emsdk_set_env.sh \
+    # Remove debugging symbols from embedded node (extra 7MB)
+    && strip -s `which node` \
+    # Tests consume ~80MB disc space
+    && rm -fr ${EMSDK}/llvm/clang/emscripten/tests \
+    # strip out symbols from clang (~extra 50MB disc space)
+    && find ${EMSDK}/llvm/clang/bin -type f -exec strip -s {} + || true \
+    && find ${EMSDK}/llvm/clang/fastcomp/bin -type f -exec strip -s {} + || true \
+&&  echo "## Done"
+
+# Populate Emscripten SDK cache with libc++, to improve further compilation times.
+RUN echo "## Pre-populate cache" \
```

Comment:
The best way to populate the cache is to run "./embuilder".     Perhaps "./embuilder build ALL`, or `./embuilder build SYSTEM` (the formor might be too slow).

Reply:
Huh this is a trace from one of the very first implementation of this image (I believe 1.34.4).
Later I was experimenting with fat build (https://github.com/trzecieu/emscripten-docker/blob/master/docker/trzeci/emscripten/Dockerfile#L55) but that solution wasn't good - once is that it has increased build time significantly, another issue was build size that wasn't acceptable anymore. 

To just pre-populate the most frequently used libs I went that way. 

I've added `embuilder.py build SYSTEM` as it looks like a sweet spot.